File: 2020/day07.cpp
#include <cstdio>
#include <iostream>
#include <map>
#include <regex>
#include <string>

using namespace std;

const regex inputFmt = regex("^([a-z]+ [a-z]+) bags contain (.*)");
const regex inputPartFmt = regex("([0-9]+) ([a-z]+ [a-z]+) bags?[,.]");

map<string, vector<pair<string, int>>> adjs;

bool find(string source, string target) {
  if(source == target) return true;
  for(auto e : adjs[source]) {
    if(find(e.first, target)) return true;
  }
  return false;
}

int count(string source) {
  int cnt = 0;
  for (auto e : adjs[source]) {
    cnt += e.second * (1 + count(e.first));
  }
  return cnt;
}

int main() {
  vector<string> types;
  string line; smatch m;
  while(getline(cin, line)) {
    if(regex_search(line, m, inputFmt)) {
      types.push_back(m.str(1));
      string parts = m.str(2);
      auto partsIt = sregex_iterator(parts.begin(), parts.end(), inputPartFmt);
      for(; partsIt != sregex_iterator(); partsIt++) {
        adjs[m.str(1)].emplace_back((*partsIt)[2], stoi((*partsIt)[1]));
      }
    }
  }
  
  int cnt1 = 0;
  for(string type : types) {
    if(type != "shiny gold") {
      cnt1 += find(type, "shiny gold");
    }
  }
  printf("%d %d\n", cnt1, count("shiny gold"));
  return 0;
}

File: 2020/day07.hs
import Control.Arrow
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Text.Regex

solve1 :: Map String [(String, Int)] -> Int
solve1 inp = subtract 1 $ length $ filter (find "shiny gold") $ Map.keys inp
  where
    find tgt src = src == tgt || any (find tgt . fst) (inp Map.! src)

solve2 :: Map String [(String, Int)] -> Int
solve2 = count "shiny gold"
  where
    count src inp = sum $ map (\(t, n) -> n + n * count t inp) (inp Map.! src)

parseInput :: String -> Map String [(String, Int)]
parseInput = Map.fromList . map readLine . lines
  where
    readLine str =
      case matchRegex (mkRegex "^([a-z]+ [a-z]+) bags contain (.*)") str of
        Just [source, parts] -> (source, readParts parts)
        _ -> error $ "Invalid line: " <> show str
    readParts str =
      case matchRegexAll (mkRegex "^([0-9]+) ([a-z]+ [a-z]+) bags?(, |.)") str of
        Just (_, _, rest, [n, part, _]) -> (part, read n) : readParts rest
        _ -> []

main :: IO ()
main = interact $ show . (solve1 &&& solve2) . parseInput
